### PR TITLE
PICARD-922: Make multi-value script functions work correctly

### PR DESCRIPTION
**Pull request by Sophist-UK**
1. Fixes $inmulti using code from #623 previously reverted by #626
2. Fixes $copymerge to deduplicate new, and retain order of old.
3. Adds a new $lenmulti function.

Resolves https://tickets.metabrainz.org/browse/PICARD-922 for Picard v2.

### CODE REVIEW

File: picard/script.py [hunk 1]
```diff
@@ -479,7 +502,14 @@ def func_copymerge(parser, new, old):
...
     oldvals = parser.context.getall(old)
-    parser.context[new] = newvals + list(set(oldvals) - set(newvals))
+    result = []
+    for x in newvals:
+        if x not in result:
+            result.append(x)
+    for x in oldvals:
+        if x not in result:
+            result.append(x)
+    parser.context[new] = result
```
Comment (zas) [inline picard/script.py]: The goal is to get each element once in the final list, right ?

If you don't mind order:

```python
parser.context[new] = list(set(oldvals + newvals))
```

If you mind order, use https://github.com/metabrainz/picard/blob/master/picard/util/__init__.py#L266 `uniqify()`

File: picard/script.py [hunk 2]
```diff
@@ -388,9 +388,32 @@ def func_in(parser, text, needle):
...
+       used to split it. In both cases, it returns true if the resulting list
+       contains exactly ``needle`` as a member."""
+
+    needle = needle.eval(parser)
+
+    if separator == MULTI_VALUED_JOINER:
+        if (isinstance(haystack, ScriptExpression) and
+                len(haystack) == 1 and
+                isinstance(haystack[0], ScriptVariable)):
+            haystack = haystack[0]
```
Comment (zas) [inline picard/script.py]: This code deserves a comment imho.

File: picard/script.py (hunk 1 above)
Comment (Sophist-UK) [inline picard/script.py]: Order needs to be preserved. So I will change this to `parser.context[new] = uniquify(newvals + oldvals)`

File: picard/script.py (hunk 2 above)
Comment (Sophist-UK) [inline picard/script.py]: Agreed.

Comment (Sophist-UK) [inline picard/script.py]: Moved into worker function with comments.

File: test/test_script.py [hunk 3]
```diff
@@ -384,3 +388,48 @@ def test_cmd_unset_multi(self):
...
+
+        context["foo"] = ["First", "Second", "Third"]
+        self.assertEqual(
+            self.parser.eval("$in(%foo%,Second)", context), "1")
+        self.assertEqual(
+            self.parser.eval("$in(%foo%,irst; Second; Thi)", context), "1")
+        self.assertEqual(
+            self.parser.eval("$in(%foo%,First; Second; Third)", context), "1")
+        self.assertEqual(
+            self.parser.eval("$inmulti(%foo%,Second)", context), "1")
```
Comment (mineo) [inline test/test_script.py]: Please add a test with a user-defined separator. At the moment, that doesn't appear to work:

```
======================================================================
ERROR: test_cmd_inmulti (test.test_script.ScriptParserTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/wieland/dev/picard/test/test_script.py", line 416, in test_cmd_inmulti
    self.parser.eval("$inmulti(%foo%,Second, :)", context), "1")
  File "/home/wieland/dev/picard/picard/script.py", line 284, in eval
    return ScriptParser._cache[key].eval(self)
  File "/home/wieland/dev/picard/picard/script.py", line 122, in eval
    result.append(item.eval(state))
  File "/home/wieland/dev/picard/picard/script.py", line 114, in eval
    return function(parser, *args)
  File "/home/wieland/dev/picard/picard/script.py", line 421, in func_inmulti
    return func_in(parser, _get_multi_values(parser, haystack, separator), needle)
  File "/home/wieland/dev/picard/picard/script.py", line 340, in _get_multi_values
    return multi.split(separator) if separator else [multi]
TypeError: must be str or None, not ScriptExpression

----------------------------------------------------------------------
Ran 61 tests in 0.033s

FAILED (errors=1)
```

**Comment by zas** [comment]
@Sophist-UK : can you fix conflicts ?

**Comment by Sophist-UK** [comment]
Yes - will look at these in an hour or so.

**Pull request merged**